File: sources/csharp/05_alloc_churn/main.cs
// task 05 alloc_churn — expected output: 1274991808
// build: coreclr: dotnet build -c Release | nativeaot: dotnet publish -c Release -p:PublishAot=true | mono: mcs -optimize+ main.cs    run: coreclr: dotnet run -c Release (or bin/Release/net8.0/main.exe) | nativeaot: bin/Release/net8.0/publish/main.exe | mono: mono main.exe
using System;

class Program
{
    static void Main()
    {
        long total = 0;
        byte[][] slots = new byte[256][];
        for (int i = 0; i < 10000000; i++)
        {
            byte[] buf = new byte[64];
            buf[0] = (byte)(i % 256);
            total += buf[0];
            // Storing the buffer keeps it reachable; the array it replaces becomes garbage.
            slots[i % 256] = buf;
        }
        Console.WriteLine(total);
    }
}

File: sources/fsharp/05_alloc_churn/main.fs
// task 05 alloc_churn — expected output: 1274991808
// build: dotnet build -c Release    run: dotnet run

let slots : byte[] array = Array.zeroCreate 256

let mutable total = 0L
let mutable i = 0
while i < 10000000 do
    let buf : byte[] = Array.zeroCreate 64
    buf.[0] <- byte (i % 256)
    total <- total + int64 buf.[0]
    // Keeps buf reachable; the buffer it replaces becomes garbage.
    slots.[i % 256] <- buf
    i <- i + 1

printfn "%d" total
